### ISSUE TITLE
Output from notebook comm messages

### ISSUE THREAD

**Issue by jasongrout** [jupyterlab/jupyterlab#1254]
While investigating https://github.com/ipython/ipywidgets/issues/44, we uncovered an interesting situation in JLab: now that widgets *are* outputs, it's not clear anymore where output should go that is generated in response to a widget comm message. In the classic notebook, we had a hack where we would find the cell associated with a view, and put the output in that cell. But sometimes the widget change is not done in a specific view (as noted in https://github.com/ipython/ipywidgets/issues/316). And in JLab, there's not a clear place to put extra output generated from another output.

In the special case of a `stdin` message from `input()`, an output can act as an input, but the input control is immediately replaced with the value.

In the case of a widget, where an output is acting as an input that needs to stay active, it's not clear where output should go. @fperez and I talked a lot about this today, and came up with the following:

* If it's important to have the output go on the page, then capture the output explicitly in an output widget and display the output widget. This is what interact does in JLab, for example.
* Otherwise, the output messages will go to the notebook and won't be claimed. 

Here is the TODO:

- [ ] Have a notebook-level log widget that can be opened from a command and maybe a toolbar button.
- [ ] Save unclaimed output messages for the most recent request (based on the parent header)
- [ ] Pop up a notification in the corner that there was output not shown on the page available in the log. 
- [ ] When the user clicks on a link in the notification, show the log widget in a new window with a log of rendered output, and append all unclaimed output to the log window. The log should separate output from different requests, perhaps with a horizontal line. 
- [ ] Notifications will no longer show up in the notebook as long as an output log is open. 
- [ ] Notifications also have an option to dismiss or dismiss all such notifications. 
- [ ] The output log also has instructions on how to capture output using the output widget. 
- [ ] Also, there might be an easy way to drag log outputs into a notebook.
- [ ] The log should be cleared when restarting the notebook? Or perhaps just when 'Clear Output' is run? Perhaps there should be a button in the log to clear it.
- [ ] Should unclaimed messages be persisted in the notebook?

**Comment by jasongrout** [comment]
Here is an example of capturing the output of a button click:

```python
import ipywidgets as widgets
button = widgets.Button(description='Display Chart')
out = widgets.Output()
def on_button_clicked(b):
    button.description = 'clicked'
    with out:
        print('Ay')

button.on_click(on_button_clicked)
widgets.VBox([button, out])
```

**Comment by ellisonbg** [comment]
Can we close this?

**Comment by jasongrout** [comment]
I think there are still lots of things we can do (see the TODO list in the original message).

**Comment by MichaelHalfon** [comment]
Hello!
I've just started to use ipywidgets and I'm interested in this issue. I was wondering, since we have to pass by the output widget to see the text, is it impossible to display graphs ? Or is there a workaround ? Right now if I create a graph in the button clicked function it doesn't show.

**Comment by mbektasbbg** [comment]
I am looking into this issue

**Comment by jasongrout** [comment]
Here is a bit more background about the decisions here for further discussion:

The notebook is an opinionated format, with clear input and output regions. The output regions are solely a side effect of the input region. When an execution is done, the output region is static and can be saved. Theoretically, an output region can be created given the kernel/world state and a single execution.

Exceptions to this simple input/output data flow, in increasing complexity:

A. Editing a cell that has been previously executed and saving without execution breaks the link between the input and output, but is not noted in the notebook. This could easily be taken care of by changing the corresponding input number once a cell was modified to indicate the input was modified since it was run.
B. The stdin iopub messages (i.e., `input()` in python). These also allow an area in the output region be an input region, but the protocol and implementation enforce the convention that execution stops until input is entered, and the raw input is immediately reflected in the output.
C. Display ids allow one cell's execution to change another cell's output. This breaks the assumption that the output saved for a cell is a result of that cell's execution.
D. ipywidgets allows interactive widgets in a cell's output to effect code execution, which may result in output.
E. ipywidgets allows inputs from physical controls that may effect executions. In this case, there is no clear place to put outputs (e.g., from a button or knob on an external hardware controller).
F. Execution from other clients may produce outputs that the notebook receives that have nothing to do with any input cells.

In the original Jupyter widgets implementation, there was a new middle area of a cell, the widget area. The cell inputs could create widgets and outputs, and the widget area acted as a second input region that could itself create outputs. In case D above, the widgets were understood to be a second input region, so it was natural to append output from widget execution in the output region of the cell. Often this output included print statements from a widget state change callback or error messages from Python. This was convenient for debugging and getting quick feedback, and made the messages local to where the user was likely focusing (the output region where the control was).

ipywidgets 7 in August 2017 did away with this specialized cell widget area, which helped widgets conform better to the existing notebook input/output structure and provided more powerful ways to build interactive GUIs intermixing outputs and inputs. However, this came at the cost of mixing the input and output areas of a cell, as noted in case D above. ipywidgets 7 provides an Output widget, which constrains these dynamic outputs to specific areas that the user can opt into. However, this was not convenient for many people and was not backwards compatible with the previous behavior, so we hacked in a way to have the case D above append output to the current cell's output region for the classic notebook.

In JupyterLab, we propose returning one step closer to the purity of outputs corresponding to inputs. Instead of case D appending outputs to the cell, outputs would be put somewhere else by default. #6833 is about putting those outputs by default in a new area, an output console. This way users can see the outputs, but they don't contaminate the output region of the cell. A side effect of this is that it also solves the problem for cases E and possibly F, which we had no solution for before.

Questions:

* Is this purity of input/output data flow worth trying to preserve?
* If so, can we make this output console convenient enough for users to
  understand and use that it fulfills their needs?

I'll also note that we explored using an actual console for this output logging. There are several limitations of the current JLab console: https://github.com/jupyterlab/jupyterlab/pull/6833#issuecomment-516979370. But even if those are fixed, a primary problem is that consoles do not store their outputs (so they have to be open in order to display outputs to the user). Also, since they have their own kernel connection, they cannot currently do things like render widgets.
